### ISSUE TITLE
Model Inference | Predictions only 

### ISSUE THREAD

**Issue by RauchLukas** [Pointcept/Pointcept#318]
Hi all, 

**In Short:** Is there a native way to perform the inference (prediction) only? 

**Explanation:** After pre-training my semantic segmentation model, I finally want to use it in a small application. 
I am not interested in the calculation of the metrics (mIoU, mAcc, etc.), but just ignoring the metric is not an option since I have no annotations for the new data.

I guess this should be the basic inference use case of these models when they are used in production. 
However, I could not find an easy way from your codebase to just run the pipeline and spit out the predictions without further evaluation. 
All I could find are the scripts for train.py and test.py. The second one has already baked the evaluation directly into the [SemSegTester](https://github.com/Pointcept/Pointcept/blob/7b37078ae301288309d62c4c88401716b6bdbe0e/pointcept/engines/test.py#L268C66-L284C81). So I can't switch it off via configs or hooks. 

But maybe I'm just stupid... 🤣🤣

Cheers, Lukas

**Comment by Gofinge** [comment]
Hey, check this test code for reference! 

```
import os
import numpy as np
import torch
import torch.nn.functional as F
from collections import OrderedDict
from pointcept.datasets import build_dataset
from pointcept.models import build_model
from pointcept.datasets.transform import Compose
from pointcept.utils.config import Config
from pointcept.utils.visualization import save_point_cloud
from pointcept.utils.comm import get_world_size
from pointcept.datasets.utils import point_collate_fn

from tensorboardX import SummaryWriter

data_config = dict(
    type="ScanNetDataset",
    split="train",
    data_root="data/scannet",
    transform=[
        dict(type="CenterShift", apply_z=True),
        dict(
            type="GridSample",
            grid_size=0.02,
            hash_type="fnv",
            mode="train",
            return_grid_coord=True,
        ),
        dict(type="CenterShift", apply_z=False),
        dict(type="NormalizeColor"),
        dict(type="ToTensor"),
        dict(
            type="Collect",
            keys=("coord", "grid_coord", "segment", "name"),
            feat_keys=("color", "normal"),
        ),
    ],
    test_mode=False,
)

model_config = dict(
    type="PT-v3m1",
    in_channels=6,
    order=("z", "z-trans", "hilbert", "hilbert-trans"),
    stride=(2, 2, 2, 2),
    enc_depths=(2, 2, 2, 6, 2),
    enc_channels=(32, 64, 128, 256, 512),
    enc_num_head=(2, 4, 8, 16, 32),
    enc_patch_size=(1024, 1024, 1024, 1024, 1024),
    dec_depths=(2, 2, 2, 2),
    dec_channels=(64, 64, 128, 256),
    dec_num_head=(4, 4, 8, 16),
    dec_patch_size=(1024, 1024, 1024, 1024),
    mlp_ratio=4,
    qkv_bias=True,
    qk_scale=None,
    attn_drop=0.0,
    proj_drop=0.0,
    drop_path=0.3,
    shuffle_orders=True,
    pre_norm=True,
    enable_rpe=False,
    enable_flash=True,
    upcast_attention=False,
    upcast_softmax=False,
    cls_mode=False,
    pdnorm_bn=False,
    pdnorm_ln=False,
    pdnorm_decouple=True,
    pdnorm_adaptive=False,
    pdnorm_affine=True,
    pdnorm_conditions=("ScanNet", "S3DIS", "Structured3D"),
)

checkpoint = "exp/scannet/semseg-pt-v3m1-0-base/model/model_last.pth"
keywords = "backbone."
replacement = ""
idx = 300

if __name__ == "__main__":
    model = build_model(model_config).cuda()
    checkpoint = torch.load(
        checkpoint, map_location=lambda storage, loc: storage.cuda()
    )
    weight = OrderedDict()
    for key, value in checkpoint["state_dict"].items():
        if not key.startswith("module."):
            key = "module." + key  # xxx.xxx -> module.xxx.xxx
        if keywords in key:
            key = key.replace(keywords, replacement)
        if get_world_size() == 1:
            key = key[7:]  # module.xxx.xxx -> xxx.xxx
        weight[key] = value
    load_state_info = model.load_state_dict(weight, strict=False)
    print(load_state_info)
    dataset = build_dataset(data_config)
    data = dataset[idx]
    for key in data.keys():
        if isinstance(data[key], torch.Tensor):
            data[key] = data[key].cuda(non_blocking=True)
    output = model(data)
    feat = output.feat
    feat = feat - feat.mean(dim=-2, keepdim=True)
    u, s, v = torch.pca_lowrank(feat, center=False, q=3)
    projection = feat @ v
    min_val = projection.min(dim=0, keepdim=True)[0]
    max_val = projection.max(dim=0, keepdim=True)[0]
    div = torch.clamp(max_val - min_val, min=1e-6)
    pca_color = (projection - min_val) / div
    os.makedirs("test/vis_pca", exist_ok=True)
    save_point_cloud(
        coord=output.coord,
        color=pca_color,
        file_path=f"test/vis_pca/pca_color{idx}.ply",
    )
    vertices = output.coord.unsqueeze(0).clone().detach().cpu().numpy() / 3
    colors = pca_color.unsqueeze(0).clone().detach().cpu().numpy() * 255

    writer = SummaryWriter("test/vis_pca")
    writer.add_mesh(
        f"train/{data['name']}",
        vertices=vertices,
        colors=colors,
        global_step=1,
    )
```

**Comment by RauchLukas** [comment]
Thank You @Gofinge. 

I think I can work with that 👍

To my understanding, what was the actual purpose of the code snipped you posted? I see that you replace the backbone with empty values (which will randomly reinitiate the weights?) 

Also, this script evaluates the features, not the final class predictions, to visualize these. Is that correct?

**Comment by Gofinge** [comment]
Yes, this is for visualization of point cloud embedding. You can modified it we some actually inference propose.

**Comment by tanmay4269** [comment]
how do i get normal features for a custom point cloud? removing it from data_config/transform/collect/feat_keys doesnt make sense coz weigths wont match then

**Comment by RauchLukas** [comment]
the above script requires your dataset to already have normal features as input for the model. 
If this is not the case, you can try to estimate the normals in advance. The dataset folder must then contain a normals.npy file for each sample in addition to the coord.npy and colour.npy files. 

In my case, the data came from a lidar scanner, which outputs no normals. Therefore I retrained the model from scratch without the normal feat_key.

**Comment by Gofinge** [comment]
Yeah, you can check my preprocessing code for how to process normal from various dataset. What’s the format of your raw data?

**Comment by tanmay4269** [comment]
My point clouds are .ply files

**Comment by Gofinge** [comment]
Hi, sorry for the late response, I just recovered from the busy schedule for the CVPR project. PlY can be mesh or point cloud, and some ply already contains normal and some are not. You can try to take a close look at the ply file to understand the format of the data. You can refer to the preprocessed code for ScanNet to see how to load a ply file
